File: LeetCodePython/py55.py
# 跳跃游戏，能否到达最后
class Solution:
    def canJump(self, nums: List[int]) -> bool:
       # 一次循环能走过最远的
        n = len(nums)
        if n<=1:
            return True
        further = 0
        for index, num in enumerate(nums):
            if index <= further and index + num > further:  # 该步可达，且在该步的基础上能到达更远，则更新可达的距离
                further = index + num
            if further >= n-1: #可达最远点已经可以到最后一个位置，则返回True
                return True
            if index > further:  # 如果该步已经不可达且还没退出，说明已经到不了了 
                return False
        return False

File: LeetCodeCPlusPlus/py55.cpp
// 跳跃游戏，能否到达最后的位置
// 不断更新可达的最远位置，直至到最后或者不可达
#include <vector>
using namespace std;
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int index = 0;
        int furthest = 0;
        while (index<=furthest) {
            furthest = furthest>index+nums[index] ? furthest: index+nums[index];
            index++;
            if (furthest >= nums.size()-1)
                return true;
        }
        return false;
    }
};